### PR TITLE
Reduce concurrency issues with web cache downloads

### PR DESCRIPTION
**Pull request by aaroncameron-wk**
#### Reason for change
Previous implementation of temporary file usage for web cache downloads was potentially problematic (concurrent downloads of the same file would conflict).
Previous implementations of other web cache file management introduced TOCTOU issues.

#### Description of change
* Use FileLock so no two WebCache instances will download the same file concurrently
* Use operations that either combine the file system check + operation or simply perform the operation and ignore the error.

#### Steps to Test
Web cache downloading isn't covered by our CI. Run two CLI commands locally that both attempt download of the same large file concurrently.

**review**:
@Arelle/arelle

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

**Review by austinmatherne-wk — APPROVED** [review]

File: arelle/WebCache.py [hunk 1]
```diff
@@ -712,40 +726,43 @@ def _downloadFile(
...
                                         level=logging.ERROR)
-                    if os.path.exists(tempFilepath):
-                        os.remove(tempFilepath)
+                    if temporaryFilename is not None:
+                        temporaryFilename.unlink(missing_ok=True)
                     return False
 
             # rename temporarily named downloaded file to desired name
-            if os.path.exists(filepath):
+            with SoftFileLock(f'{filepath}.lock'):
```
Comment (austinmatherne-wk) [inline arelle/WebCache.py:734]: Add a timeout so we don't potentially hang forever if Arelle crashes on a previous execution?

**Review by austinmatherne-wk — APPROVED** [review]

File: arelle/WebCache.py [hunk 2]
```diff
@@ -208,6 +216,11 @@ def httpsRedirect(self):
     def httpsRedirect(self, value):
         self._httpsRedirect = value
 
+    def __cleanupLocks(self):
+        if os.path.exists(self.cacheDir):
+            for filename in glob.iglob(f'{self.cacheDir}/**/*.lock', recursive=True):
```
[the remark this answers is not in the document]
Comment (brettkail-wk) [reply arelle/WebCache.py:221]: It seems like deleting the file breaks the locking, or do I misunderstand the mechanism?
- process1 starts and deletes all lock files
- process1 creates a lock file and starts downloading
- process2 starts and deletes all lock files, including the one process1 is using?
- process2 creates a lock file and starts downloading again?

It seems like we could just create the file if it doesn't exist and then attempt to lock it without checking ctime at all.

File: arelle/WebCache.py [hunk 3]
```diff
@@ -712,40 +727,43 @@ def _downloadFile(
...
                                         level=logging.ERROR)
-                    if os.path.exists(tempFilepath):
-                        os.remove(tempFilepath)
+                    if temporaryFilename is not None:
+                        temporaryFilename.unlink(missing_ok=True)
                     return False
 
             # rename temporarily named downloaded file to desired name
-            if os.path.exists(filepath):
+            with SoftFileLock(f'{filepath}.lock', timeout=FILE_LOCK_TIMEOUT) as lock:
```
[the remark this answers is not in the document]
Comment (brettkail-wk) [reply arelle/WebCache.py:735]: That approach would also mean that we don't end up with multiple half-downloaded files.  For example, user attempts to download a large file to a random filename, but it takes too long so the user kills the process after downloading several hundred MB and tries again.  The second process does the same, but the user kills it again, so they now have two randomly named files with several hundred MB.  The third attempt succeeds, but nothing cleans up the randomly named files, just the empty lock files.

File: arelle/WebCache.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aaroncameron-wk) [reply arelle/WebCache.py:221]: I initially misinterpreted the `filelock` docs which led me to choosing `SoftFileLock`, but looking again `FileLock` will work for us and will remove the risk of lingering lock files and so no cleanup is necessary.

[the remark this answers is not in the document]
Comment (brettkail-wk) [reply arelle/WebCache.py:221]: > If Arelle exits uncleanly and leaves a .tmp file it's not using, we won't want to be blocked from downloading indefinitely

File locks are owned by a process, so if the process exits uncleanly, the lock will be automatically released.  With that, I don't think we need ctime checking at all?

File: requirements.txt [hunk 4]
```diff
@@ -1,5 +1,6 @@
 # core modules
 certifi==2024.6.2
+filelock==3.15.1
```
Comment (austinmatherne-wk) [inline requirements.txt:3]: This also needs to be in `pyproject.toml` for the Python library.

**Review by allenbross-wf — APPROVED** [review]

**Review by austinmatherne-wk — APPROVED** [review]

**Comment by allenbross-wf** [comment]
Verified against GitHub issue.
QA +10

**Pull request merged**